### DISCUSSION TITLE
#412 is not answered

### DISCUSSION THREAD

**Discussion by awwright** [Homebrew/discussions#903] [Everyday usage]
I filed discussion #412 because I had what I thought was a straightforward question: Apple promised that building Universal binaries would be super duper simple, and I just wanted to know what sort of roadblocks occurred in practice.

I'm very sorry that the thread was derailed so easily, but I don't feel that's my fault.

It shouldn't be locked, and it certainly shouldn't be marked as answered.

**Comment by SMillerDev** [comment]
If you want to know how people use apple technology, ask stackoverflow.

The question "why doesn't brew do this?" was answered. It's because it isn't worth the effort.

**Comment by danielnachun** [comment]
I'm not sure what additional answers you would like us to give.  Cross-compiling is very difficult and we have a solution which lets us avoid that.  Building universal binaries may indeed be easy if you are distributing software use `.pkg` files with installers which package all their dependencies not provided by the system.  But that is not what Homebrew is designed to do.  

It's also worth reminding everyone that we currently have to support 4 operating systems - macOS 10.14, macOS 10.15, macOS 11.0 and Linux, and a huge number of build systems and programming languages.  If you look through most maintainers' commit history, you will see how much effort it already takes just for that to work.  Cross-compiling will just make things that much more difficult.

**Reply by tjluoma** [reply]
> It's also worth reminding everyone that we currently have to support 4 operating systems
> macOS 10.14, macOS 10.15, macOS 11.0 and Linux,

Somewhat related to this… I'm not sure when the addition of Linux to Homebrew happened, and I'm sure there were compelling reasons to do so, but I have tried googling to find out "Why did Homebrew decide to expand the Linux?" and have not really found anything (lots of articles about it happening, but not about the background/decision-making).

Rather than ask the same questions (which are now irrelevant since the decision has been made), I wondered if someone could point me to something that would explain how that decision came to be.

I ask here because it's clear that the addition of Linux has "opportunity costs" (not necessarily for the issue of universal binaries for Mac, but definitely some) so I'd like to understand what made that a compelling idea to undertake.

**Reply by danielnachun** [reply]
@sjackman initiated the addition of Linux and may have additional thoughts.  I think the biggest reason was that Homebrew's architecture was able to accommodate adding Linux without major alterations.  In many ways supporting Linux is actually less complex than macOS.  The only parts of the host system we use are `gcc` and `glibc` (and even these are brewed if the host systems versions are too old), and virtualization for VMs to run CI is far more flexible than on macOS.  

By contrast, Homebrew on macOS uses a lot more of the system libraries, and getting CI set up for macOS is more niche.  In my experience, supporting multiple versions of macOS is a lot harder in many cases than Linux, because there are sometimes substantial differences between versions.  

Although I've only been contributing to Homebrew for a few months, from my experience so far I don't really see cases where we are held back by needing to get things to work on Linux.  I've got more ideas in the works to improve our Linux support, but in many cases those improvements will actually benefit Homebrew on macOS as well.

**Reply by SMillerDev** [reply]
> I'm not sure when the addition of Linux to Homebrew happened, and I'm sure there were compelling reasons to do so, but I have tried googling to find out "Why did Homebrew decide to expand the Linux?" and have not really found anything (lots of articles about it happening, but not about the background/decision-making).

The official support was added in homebrew 2.0.0. It was added simply because someone thought it was useful and put in the work to make it happen. There was no grand plan, there rarely is in homebrew, just someone putting in the effort until something they found useful worked.

I'm sure Shaun can explain better why it was useful to him, but as I understand it there is a lot of people in the scientific computing world who have limited options to use system package managers so they would need to compile software for themselves. Homebrew makes that a lot easier.

**Reply by tjluoma** [reply]
It was my understanding that Linux had a lot of different package-managers already, but I had never really used Linux. Upon a little further study, it appears that there are several different package-manages on Linux, but they aren't all available on every version/distro of Linux, and some of them are not very well maintained or have other drawbacks. Plus they all seem to require `sudo` which "Linuxbrew" does not. So that makes a lot more sense already.

**Reply by danielnachun** [reply]
The only options that don't require `sudo` either build everything from source (i.e. Gentoo Prefix), use the unmaintained and very slow Proot to do a `chroot` without `sudo` (i.e. Junest, Nix), or require the use of virtual environments (i.e. Anaconda).  Homebrew is the only Linux package manager that 'just works' without sudo.  We still have to build some packages from source but I'm hoping to port the binary patching used by Anaconda so that Homebrew on Linux can be a fully binary package manager.   This would make it as easy to use as the traditional package managers on Linux (apt, dpkg, etc.) but able to run from any directory without the need for sudo.  This has been a "pipe dream" for years in the Linux world but I actually think we are very close to achieving it here.  Hopefully that provides some context of why I'm excited about Linux support in Homebrew!

**Reply by sjackman** [reply]
> - Can install software to your home directory and so does not require sudo
> - Install software not packaged by your host distribution
> - Install up-to-date versions of software when your host distribution is old
> - Use the same package manager to manage your macOS, Linux, and Windows systems

See https://docs.brew.sh/Homebrew-on-Linux#features
